pinot-core/src/main/java/org/apache/pinot/server/realtime/ControllerLeaderLocator.java

```diff
@@ -80,26 +81,27 @@ public static ControllerLeaderLocator getInstance() {
 
   /**
    * Locates the controller leader so that we can send LLC segment completion requests to it.
-   * Checks the {@link ControllerLeaderLocator::_cachedControllerLeaderInvalid} flag and fetches the leader from helix if cached value is invalid
+   * Checks the {@link ControllerLeaderLocator::_cachedValidControllerLeaderMap} and fetches the leader from helix if cached value is invalid or missing.
    * @param rawTableName table name without type.
    * @return The host-port pair of the current controller leader.
    */
   public synchronized Pair<String, Integer> getControllerLeader(String rawTableName) {
-    if (!_cachedControllerLeaderInvalid) {
-      return _controllerLeaderHostPort;
+    Pair<String, Integer> leaderPairForTable = _cachedValidControllerLeaderMap.get(rawTableName);
```

Comment:
I suppose we expect that the number of tables having consuming segments in a host is generally smaller than the number of partitions in the lead controller resource.  If it is the other way around then it will nbe better to cache the partitionId to leader pair. I am fine with this approach for now, but we should at least add a comment that we are taking that approach.

Reply:
Good point. Updated.